walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
+    int meetingDuration = (int) request.getDuration();
+
+    // check for no attendees
+    if (attendeesRequested.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // check duration of meeting request
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+
+    List<TimeRange> badTimes = getBadTimes(events, attendeesRequested);
+
+    // list to hold all the available times (everything not in badTimes)
+    List<TimeRange> goodTimes = getGoodTimes(badTimes, meetingDuration);
+    return goodTimes;
+  }
+
+  /** Returns a List with all the times that requested attendees have conflicts. */
+  private List<TimeRange> getBadTimes(
+      Collection<Event> events, Collection<String> attendeesRequested) {
+    List<TimeRange> badTimes = new ArrayList<TimeRange>();
+
+    // iterate through all events and add  time to badTimes if attendee is in attendeesRequested
+    for (Event conflictEvent : events) {
+      Set<String> busyAttendees = conflictEvent.getAttendees();
+
+      // loop to see if any attendees in conflictEvent are in attendeesRequested
+      Boolean conflict = false;
+      Iterator<String> eventIter = busyAttendees.iterator();
+      while (!conflict && eventIter.hasNext()) {
+        if (attendeesRequested.contains(eventIter.next())) {
+          conflict = true;
+        }
+      }
+      if (conflict == true) { // there was an attendee with a conflict, so add bad time to List
+        badTimes.add(conflictEvent.getWhen());
+      }
+    }
+
+    // sort badTimes
+    Collections.sort(badTimes, TimeRange.ORDER_BY_START);
+    return badTimes;
+  }
+
+  /**
+   * Returns a List with all the TimeRanges between the badTimes that are greater or equal to the
+   * requested meeting duration.
+   */
+  private List<TimeRange> getGoodTimes(List<TimeRange> badTimes, int meetingDuration) {
+    List<TimeRange> goodTimes = new ArrayList<TimeRange>();
+
+    // beginning of an available period
+    int goodStart = TimeRange.START_OF_DAY;
+
+    // end of an available period
+    int goodEnd = TimeRange.END_OF_DAY;
+
+    for (TimeRange badRange : badTimes) {
+      goodEnd = badRange.start();
+
+      if (goodEnd > goodStart) {
+        TimeRange goodRange = TimeRange.fromStartEnd(goodStart, goodEnd, false);
+        if (goodRange.duration() >= meetingDuration) {
+          goodTimes.add(goodRange);
+        }
+      }
+
+      // only increment goodStart if it is before the next ending period - takes care of nested
+      // events
+      if (goodStart < (goodEnd + badRange.duration())) {
+        goodStart = goodEnd + badRange.duration();
+      }
+    }
+
+    // edge case for last time chunk
+    if (goodStart < TimeRange.END_OF_DAY) {
+      TimeRange goodRange = TimeRange.fromStartEnd(goodStart, TimeRange.END_OF_DAY, true);
+      if (goodRange.duration() >= meetingDuration) {
+        goodTimes.add(goodRange);
+      }
+    }
```

Comment:
Slightly more succinct, but entirely stylistic (leaving it at your discretion, no action required):
if (TimeRange.END_OF_DAY - goodStart > meetingDuration) {
  goodTimes.add(TimeRange.fromStartEnd(goodStart, TimeRange.END_OF_DAY, true));
}
Also works above.

Reply:
I like this better, thanks!